### PR TITLE
3rmbdDO4 Added in Pundit and defined user roles in code

### PR DESCRIPTION
**Pull request by j-maynard**
Adds in Pundit.  Creates a new User Roles Permission class which maps user roles to permissions in code.

### CODE REVIEW

File: app/controllers/msa_components_controller.rb [hunk 1]
```diff
@@ -4,20 +4,32 @@ def index
   end
 
   def new
-    @component = NewMsaComponentEvent.new
+    begin
+      authorize NewMsaComponentEvent
```
Comment (jakubmiarka) [inline app/controllers/msa_components_controller.rb]: this should not be part of this PR/ticket... and this logic would ideally be part of the `before_` callback instead of polluting the method code itself...

File: app/models/remote_authenticatable.rb [hunk 2]
```diff
@@ -72,8 +72,8 @@ def auth_complete(resp, params)
         self.email = user_attributes['email']
         self.phone_number = user_attributes['phone_number']
         self.access_token = access_token
-        self.organisation = user_attributes['custom:organisation']
```
Comment (jakubmiarka) [inline app/models/remote_authenticatable.rb]: thanks for cleaning this up

File: spec/controllers/msa_components_controller_spec.rb [hunk 3]
```diff
@@ -12,4 +12,18 @@
       expect(response).to have_http_status(:success)
     end
   end
+
+  describe "GET #new" do
+    it "returns http redirect with other users" do
```
Comment (jakubmiarka) [inline spec/controllers/msa_components_controller_spec.rb]: not sure what `... with other users` means?

File: spec/controllers/msa_components_controller_spec.rb [hunk 4]
```diff
@@ -12,4 +12,18 @@
...
+      stub_auth
+      get :new
+      expect(response).to have_http_status(:redirect)
```
Comment (jakubmiarka) [inline spec/controllers/msa_components_controller_spec.rb]: if this is testing the unauthorized flow, might be worth to have an expectation for the flash error message (and it would also made the test clearer)

File: spec/controllers/sp_components_controller_spec.rb [hunk 5]
```diff
@@ -2,15 +2,26 @@
...
+
+  describe "GET #new" do
+    it "returns http redirect with other users" do
+      stub_auth
+      get :new
+      expect(response).to have_http_status(:redirect)
+    end
+
+    it "returns http success with component manager user" do
+      certmgr_stub_auth
```
Comment (jakubmiarka) [inline spec/controllers/sp_components_controller_spec.rb]: the test name talks about `manager user`, but we're stubbing `certmgr`

File: spec/factories/users.rb [hunk 6]
```diff
@@ -4,5 +4,15 @@
...
     password { "validpassword" }
+    permissions { UserRolePermissions.new("", nil) }
+  end
+
+  factory :component_user, class: User do
+    given_name { "Jane" }
+    family_name  { "Smith" }
+    email { "test@test.test" }
+    password { "validpassword" }
+    roles { "certmgr" }
```
Comment (jakubmiarka) [inline spec/factories/users.rb]: not sure about the naming, we're giving the user the `certmgr` role but we call the factory `:component_user`?

File: spec/support/auth_support.rb [hunk 7]
```diff
@@ -1,9 +1,20 @@
...
+    allow(controller).to receive(:current_user).and_return(user)
+    sign_in(user, scope: :user)
+  end
+
+  def certmgr_stub_auth
+    @request.env["devise.mapping"] = Devise.mappings[:user]
+    user = FactoryBot.create(:component_user)
+    allow(request.env['warden']).to receive(:authenticate!).and_return(user)
+    allow(controller).to receive(:current_user).and_return(user)
+    sign_in(user, scope: :user)
```
Comment (jakubmiarka) [inline spec/support/auth_support.rb]: these 3 lines are repeated in each method, we could extract them to reduce some of the code duplication...

File: app/controllers/msa_components_controller.rb [hunk 8]
```diff
@@ -4,20 +4,32 @@ def index
...
+      @component = NewMsaComponentEvent.new
+    rescue Pundit::NotAuthorizedError
+      flash[:warn] = "You are not authorised to do this action"
```
Comment (jakubmiarka) [inline app/controllers/msa_components_controller.rb]: any content/text strings should be now defined in the locale file

File: app/models/user.rb [hunk 9]
```diff
@@ -20,4 +22,16 @@ class User
   def initialize(options = {}); end
 
   def save!; end
+
+  def attributes
```
Comment (jakubmiarka) [inline app/models/user.rb]: is this devise method? I believe we could then just do
```
def attributes
    super.merge({roles: nil, permissions: nil})
end
```
if that's all we're doing here

File: app/controllers/msa_components_controller.rb (hunk 8 above)
Comment (Rosa-Fox) [reply app/controllers/msa_components_controller.rb]: have updated!
